tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -325,6 +326,130 @@ class HLSLCallDiagnoseVisitor
     return true;
   }
 
+  bool VisitDeclRefExpr(DeclRefExpr *DRE) {
+    // Search for enum types that should only exist after
+    // specific shader models.
+
+    // check if this declrefexpr refers to another one
+    // and drill into it if possible.
+    VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl());
+    if (!VD) {
+      return true;
+    }
+
+    ImplicitCastExpr *ICE = nullptr;
+    if (VD->getInit()) {
+      if ((ICE = dyn_cast<ImplicitCastExpr>(VD->getInit()))) {
+        while (true) {
+          ImplicitCastExpr *NewICE =
+              dyn_cast<ImplicitCastExpr>(ICE->getSubExpr());
+          if (!(NewICE)) {
+            break;
+          }
+          ICE = NewICE;
+        }
+      }
+    }
+
+    if (ICE) {
+      if (DeclRefExpr *newDRE = dyn_cast<DeclRefExpr>(ICE->getSubExpr())) {
+        VisitDeclRefExpr(newDRE);
+      }
+    }
+
+    if (!DRE->getDecl()->hasAttr<AvailabilityAttr>())
+      return true;
+
+    // Next step: given a valid DRE, get an availability
+    // attribute from it
+    AvailabilityAttr *AAttr = nullptr;
+
+    AAttr = DRE->getDecl()->getAttr<AvailabilityAttr>();
+    VersionTuple AAttrVT = AAttr->getIntroduced();
+    VersionTuple SMVT = VersionTuple(SM->GetMajor(), SM->GetMinor());
+
+    // if the current shader model is lower than what
+    // is stated in the availability attribute, emit
+    // the warning warn_hlsl_builtin_constant_unavailable
+
+    if (SMVT < AAttrVT) {
+      IdentifierInfo *II = DRE->getDecl()->getIdentifier();
+      sema->Diag(DRE->getLocation(),
+                 diag::warn_hlsl_builtin_constant_unavailable)
+          << II->getName() << SM->GetName() << AAttrVT.getAsString();
+    }
+
+    return true;
+  }
+
+  bool IsRayFlagForceOMM2StateSet(const CallExpr *CE) {
+    const Expr *Expr1 = CE->getArg(1);
+    llvm::APSInt constantResult;
+    return Expr1->isIntegerConstantExpr(constantResult,
+                                        sema->getASTContext()) &&
+           (constantResult.getLimitedValue() &
+            (uint64_t)DXIL::RayFlag::ForceOMM2State) != 0;
+  }
+
+  void DiagnoseTraceRayInline(const MemberExpr *ME,
+                              CXXMemberCallExpr *callExpr) {
+    // Validate if the RayFlag parameter has RAY_FLAG_FORCE_OMM_2_STATE set,
+    // the RayQuery decl must have RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS set,
+    // otherwise emit a diagnostic.
+    bool IsRayFlagForceOMM2State = IsRayFlagForceOMM2StateSet(callExpr);
+    if (IsRayFlagForceOMM2State) {
```

Comment:
No point keeping a variable around if it's only used in the if expression.
```suggestion
    if (IsRayFlagForceOMM2StateSet(callExpr)) {
```